src/parser/toplevel.lisp

```diff
@@ -1619,91 +1619,61 @@ consume all attributes")))
-                   :type :secondary
-                   :span (cst:source (cst:second form))
-                   :message "in this class definition")))))
-
-  ;; (m d t ...)
-  (unless (or (cst:null (cst:rest (cst:rest method-form)))
-              (cst:null (cst:rest (cst:rest (cst:rest method-form)))))
-    (error 'parse-error
-           :err (coalton-error
-                 :span (cst:source (cst:first (cst:rest (cst:rest method-form))))
-                 :file file
-                 :message "Malformed method definition"
-                 :primary-note "unexpected trailing form"
-                 :notes
-                 (list
-                  (make-coalton-error-note
-                   :type :secondary
-                   :span (cst:source (cst:second form))
-                   :message "in this class definition")))))
-
-  ;; (0.5 t ...)
-  (unless (and (cst:atom (cst:first method-form))
-               (identifierp (cst:raw (cst:first method-form))))
-    (error 'parse-error
-           :err (coalton-error
-                 :span (cst:source (cst:first method-form))
-                 :file file
-                 :message "Malformed method definition"
-                 :primary-note "expected symbol"
-                 :notes
-                 (list
-                  (make-coalton-error-note
-                   :type :secondary
-                   :span (cst:source (cst:second form))
-                   :message "in this class definition")))))
-
-  ;; either list of length 2 or list of length 3 with docstring
-  (unless (or (cst:null (cst:rest (cst:rest method-form)))
-              (and (cst:atom (cst:second method-form))
-                   (stringp (cst:raw (cst:second method-form)))))
-    (error 'parse-error
-           :err (coalton-error
-                 :span (cst:source (cst:second method-form))
-                 :file file
-                 :message "Malformed method definition"
-                 :primary-note "expected docstring"
-                 :notes
-                 (list
-                  (make-coalton-error-note
-                   :type :secondary
-                   :span (cst:source (cst:second form))
-                   :message "in this class definition")))))
-
-  (let (docstring)
-    (when (and (cst:atom (cst:second method-form))
-               (stringp (cst:raw (cst:second method-form))))
-      (setf docstring (cst:raw (cst:second method-form))))
-
-    (make-method-definition
-     :name (make-identifier-src
-            :name (node-variable-name (parse-variable (cst:first method-form) file))
-            :source (cst:source (cst:first method-form)))
-     :docstring docstring
-     :type (parse-qualified-type (if docstring
-                                     (cst:third method-form)
-                                     (cst:second method-form))
-                                 file)
-     :source (cst:source method-form))))
+  (macrolet ((fail (local-form note)
```

Comment:
Are these changes just reorganizational or are they changing behavior?

Reply:
These are purely reorganizational, though they do drop some repeated probing for nullity to check the length of the method form.

Reply:
If it's just reorganizational, I wonder if that should be a separate PR applying the same structure to each parse function? I think it's reasonable to find a more condensed structure but it should probably be one fell swoop.

Reply:
I recall what the motivation was: reporting 'unexpected trailing form' needs to consider whether or not a docstring is present in order to decide whether to make the third or fourth form. The original version marks the wrong one:

```
Unhandled FIASCO::FAILED-ASSERTION in thread #<SB-THREAD:THREAD tid=259 "main thread" RUNNING
                                                {7004FD06F3}>:
  Test assertion failed when running TEST-PARSER:

Strings differ at offset 49 of expected error define-class.21.error (A) and generated error (B):
A: error: Malformed method definition
  --> test:5:13
   |
 4 |  (define-class (C :a)
   |                ------ in this class definition
 5 |   (m "doc" :a :b))
   |               ^^ unexpected trailing form

B: error: Malformed method definition
  --> test:5:10
   |
 4 |  (define-class (C :a)
   |                ------ in this class definition
 5 |   (m "doc" :a :b))
   |            ^^ unexpected trailing form
```

(also, the method is 40 ll shorter)

Reply:
Potential minimal patch:
```
1 file changed, 20 insertions(+), 18 deletions(-)
src/parser/toplevel.lisp | 38 ++++++++++++++++++++------------------

modified   src/parser/toplevel.lisp
@@ -1645,7 +1645,7 @@ consume all attributes")))
               (cst:null (cst:rest (cst:rest (cst:rest method-form)))))
     (error 'parse-error
            :err (coalton-error
-                 :span (cst:source (cst:first (cst:rest (cst:rest method-form))))
+                 :span (cst:source (cst:first (cst:rest (cst:rest (cst:rest method-form)))))
                  :file file
                  :message "Malformed method definition"
                  :primary-note "unexpected trailing form"
@@ -1672,28 +1672,30 @@ consume all attributes")))
                    :span (cst:source (cst:second form))
                    :message "in this class definition")))))
 
-  ;; either list of length 2 or list of length 3 with docstring
-  (unless (or (cst:null (cst:rest (cst:rest method-form)))
-              (and (cst:atom (cst:second method-form))
-                   (stringp (cst:raw (cst:second method-form)))))
-    (error 'parse-error
-           :err (coalton-error
-                 :span (cst:source (cst:second method-form))
-                 :file file
-                 :message "Malformed method definition"
-                 :primary-note "expected docstring"
-                 :notes
-                 (list
-                  (make-coalton-error-note
-                   :type :secondary
-                   :span (cst:source (cst:second form))
-                   :message "in this class definition")))))
-
   (let (docstring)
     (when (and (cst:atom (cst:second method-form))
                (stringp (cst:raw (cst:second method-form))))
       (setf docstring (cst:raw (cst:second method-form))))
 
+    ;; either list of length 2 or list of length 3 with docstring
+    (unless (or (cst:null (cst:rest (cst:rest method-form)))
+                (and (cst:atom (cst:second method-form))
+                     (stringp (cst:raw (cst:second method-form)))))
+      (error 'parse-error
+             :err (coalton-error
+                   :span (cst:source (if docstring
+                                         (cst:fourth method-form)
+                                         (cst:third method-form)))
+                   :file file
+                   :message "Malformed method definition"
+                   :primary-note "unexpected trailing form"
+                   :notes
+                   (list
+                    (make-coalton-error-note
+                     :type :secondary
+                     :span (cst:source (cst:second form))
+                     :message "in this class definition")))))
+
     (make-method-definition
      :name (make-identifier-src
             :name (node-variable-name (parse-variable (cst:first method-form) file))


```

Reply:
In the name of getting tests back into CI, the most straightforward thing to do seem to be to postpone any changes here and update the define-class.4 parser test to reflect current behavior of docstring parsing -- I've done that.